File: packages/core-ts/src/zipsafe.ts
import { createWriteStream, promises as fs } from 'node:fs';
import { dirname, resolve, sep } from 'node:path';
import { pipeline } from 'node:stream/promises';
import { Transform } from 'node:stream';
import yauzl from 'yauzl';
import { BridgeError } from './types.js';

export const ZIP_MAX_ENTRIES = 500;
export const ZIP_MAX_TOTAL_BYTES = 1024 * 1024 * 1024;

/**
 * 条目名安全性：拒绝绝对路径、反斜杠开头、盘符、`..` 段、NUL、空名；反斜杠统一视为分隔符（Windows 打的包）。
 * 返回规范化后的相对路径（正斜杠），不安全返回 null。
 */
export function sanitizeZipEntryName(name: string): string | null {
  if (!name || name.includes('\x00')) return null;
  if (name.startsWith('/') || name.startsWith('\\')) return null;
  if (/^[A-Za-z]:/.test(name)) return null;
  const segments = name.replace(/\\/g, '/').split('/');
  for (const seg of segments) {
    if (seg === '..') return null;
  }
  const cleaned = segments.filter((s) => s !== '' && s !== '.').join('/');
  return cleaned || null;
}

/** ZIP 外部属性高 16 位是 Unix mode；S_IFLNK 表示符号链接，整包拒绝。 */
export function isSymlinkEntry(externalFileAttributes: number): boolean {
  return ((externalFileAttributes >>> 16) & 0xf000) === 0xa000;
}

/**
 * 流式安全解压到 destDir：条目名校验 + realpath 前缀断言 + 条目数 / 总量限额；解压时按实际字节计数（不信头里的 uncompressedSize）。
 * 任一失败会删掉整个 destDir 后抛 BridgeError。
 */
export async function extractZipSafe(
  zipPath: string,
  destDir: string,
  opts: { maxEntries?: number; maxTotalBytes?: number } = {},
): Promise<void> {
  const maxEntries = opts.maxEntries ?? ZIP_MAX_ENTRIES;
  const maxTotal = opts.maxTotalBytes ?? ZIP_MAX_TOTAL_BYTES;
  await fs.mkdir(destDir, { recursive: true });
  const destReal = await fs.realpath(destDir);
  const prefix = destReal.endsWith(sep) ? destReal : destReal + sep;
  try {
    await new Promise<void>((resolveDone, reject) => {
      // decodeStrings:false 让 yauzl 不做自己的路径校验，把原始名交给我们统一判定（这样 sanitize 被注掉时夹具才会变红）
      yauzl.open(zipPath, { lazyEntries: true, decodeStrings: false, autoClose: true }, (err, zipfile) => {
        if (err || !zipfile) return reject(err ?? new Error('yauzl open failed'));
        let count = 0;
        let total = 0;
        const fail = (e: unknown) => {
          zipfile.close();
          reject(e);
        };
        zipfile.on('error', fail);
        zipfile.on('end', () => resolveDone());
        zipfile.on('entry', (entry: yauzl.Entry) => {
          void (async () => {
            count += 1;
            if (count > maxEntries) throw new BridgeError('ZIP_TOO_MANY_ENTRIES', `zip has more than ${maxEntries} entries`);
            const rawName = Buffer.isBuffer(entry.fileName) ? (entry.fileName as unknown as Buffer).toString('utf8') : String(entry.fileName);
            const rel = sanitizeZipEntryName(rawName);
            if (rel === null) throw new BridgeError('ZIP_UNSAFE_ENTRY', `unsafe zip entry: ${rawName}`);
            if (isSymlinkEntry(entry.externalFileAttributes)) throw new BridgeError('ZIP_UNSAFE_ENTRY', `symlink zip entry: ${rawName}`);
            const target = resolve(destReal, rel);
            if (!target.startsWith(prefix)) throw new BridgeError('ZIP_UNSAFE_ENTRY', `zip entry escapes target dir: ${rawName}`);
            const isDir = rawName.endsWith('/') || rawName.endsWith('\\');
            if (isDir) {
              await fs.mkdir(target, { recursive: true });
              zipfile.readEntry();
              return;
            }
            if (total + entry.uncompressedSize > maxTotal) throw new BridgeError('ZIP_TOO_LARGE', `zip expands beyond ${maxTotal} bytes`);
            await fs.mkdir(dirname(target), { recursive: true });
            // 写之前再确认父目录 realpath 仍在 dest 内（符号链接已整包拒绝，这是双保险）
            const parentReal = await fs.realpath(dirname(target));
            if (parentReal !== destReal && !parentReal.startsWith(prefix)) {
              throw new BridgeError('ZIP_UNSAFE_ENTRY', `zip entry parent escapes target dir: ${rawName}`);
            }
            const stream = await new Promise<NodeJS.ReadableStream>((res, rej) =>
              zipfile.openReadStream(entry, (e, s) => (e || !s ? rej(e ?? new Error('openReadStream failed')) : res(s))),
            );
            const counter = new Transform({
              transform(chunk: Buffer, _enc, cb) {
                total += chunk.length;
                if (total > maxTotal) return cb(new BridgeError('ZIP_TOO_LARGE', `zip expands beyond ${maxTotal} bytes`));
                cb(null, chunk);
              },
            });
            await pipeline(stream, counter, createWriteStream(target));
            zipfile.readEntry();
          })().catch(fail);
        });
        zipfile.readEntry();
      });
    });
  } catch (err) {
    await fs.rm(destDir, { recursive: true, force: true });
    if (err instanceof BridgeError) throw err;
    throw new BridgeError('ZIP_UNSAFE_ENTRY', `zip could not be read: ${(err as Error).message}`);
  }
}

File: packages/unity/Editor/Core/ZipSafe.cs
using System;
using System.IO;
using System.IO.Compression;

namespace RezonaLab.EngineBridge.Editor
{
    /// <summary>
    /// 安全解压：条目名校验 + GetFullPath 前缀断言 + 条目数 / 总量限额；解压时按实际字节计数（不信头里的 uncompressedSize）。
    /// 任一失败会删掉整个 destDir 后抛 BridgeException。
    /// </summary>
    public static class ZipSafe
    {
        public const int MaxEntries = 500;
        public const long MaxTotalBytes = 1024L * 1024 * 1024;

        /// <summary>
        /// 条目名安全性：拒绝绝对路径、反斜杠开头、盘符、`..` 段、NUL、空名；反斜杠统一视为分隔符（Windows 打的包）。
        /// 返回规范化后的相对路径（正斜杠），不安全返回 null。
        /// </summary>
        public static string SanitizeEntryName(string name)
        {
            if (string.IsNullOrEmpty(name) || name.IndexOf('\0') >= 0) return null;
            if (name[0] == '/' || name[0] == '\\') return null;
            if (name.Length >= 2 && char.IsLetter(name[0]) && name[1] == ':') return null;
            var segments = name.Replace('\\', '/').Split('/');
            var kept = new System.Collections.Generic.List<string>();
            foreach (var seg in segments)
            {
                if (seg == "..") return null;
                if (seg.Length > 0 && seg != ".") kept.Add(seg);
            }
            return kept.Count == 0 ? null : string.Join("/", kept);
        }

        /// <summary>ZIP 外部属性高 16 位是 Unix mode；S_IFLNK 表示符号链接，整包拒绝。</summary>
        public static bool IsSymlink(int externalAttributes)
        {
            return (((uint)externalAttributes >> 16) & 0xF000) == 0xA000;
        }

        public static void Extract(string zipPath, string destDir, int maxEntries = MaxEntries, long maxTotalBytes = MaxTotalBytes)
        {
            Directory.CreateDirectory(destDir);
            var destFull = Path.GetFullPath(destDir).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
            var prefix = destFull + Path.DirectorySeparatorChar;
            try
            {
                using (var archive = ZipFile.OpenRead(zipPath))
                {
                    var count = 0;
                    long total = 0;
                    foreach (var entry in archive.Entries)
                    {
                        count += 1;
                        if (count > maxEntries) throw new BridgeException("ZIP_TOO_MANY_ENTRIES", "zip has more than " + maxEntries + " entries");
                        var rawName = entry.FullName;
                        var rel = SanitizeEntryName(rawName);
                        if (rel == null) throw new BridgeException("ZIP_UNSAFE_ENTRY", "unsafe zip entry: " + rawName);
                        if (IsSymlink(entry.ExternalAttributes)) throw new BridgeException("ZIP_UNSAFE_ENTRY", "symlink zip entry: " + rawName);
                        var target = Path.GetFullPath(Path.Combine(destFull, rel.Replace('/', Path.DirectorySeparatorChar)));
                        if (!target.StartsWith(prefix, StringComparison.Ordinal)) throw new BridgeException("ZIP_UNSAFE_ENTRY", "zip entry escapes target dir: " + rawName);
                        var isDir = rawName.EndsWith("/", StringComparison.Ordinal) || rawName.EndsWith("\\", StringComparison.Ordinal);
                        if (isDir)
                        {
                            Directory.CreateDirectory(target);
                            continue;
                        }
                        if (total + entry.Length > maxTotalBytes) throw new BridgeException("ZIP_TOO_LARGE", "zip expands beyond " + maxTotalBytes + " bytes");
                        var parent = Path.GetDirectoryName(target);
                        Directory.CreateDirectory(parent);
                        // 写之前再确认父目录仍在 dest 内（符号链接已整包拒绝，这是双保险）
                        var parentFull = Path.GetFullPath(parent);
                        if (parentFull != destFull && !parentFull.StartsWith(prefix, StringComparison.Ordinal))
                            throw new BridgeException("ZIP_UNSAFE_ENTRY", "zip entry parent escapes target dir: " + rawName);
                        using (var input = entry.Open())
                        using (var output = new FileStream(target, FileMode.CreateNew, FileAccess.Write, FileShare.None))
                        {
                            var buf = new byte[64 * 1024];
                            int n;
                            while ((n = input.Read(buf, 0, buf.Length)) > 0)
                            {
                                total += n;
                                if (total > maxTotalBytes) throw new BridgeException("ZIP_TOO_LARGE", "zip expands beyond " + maxTotalBytes + " bytes");
                                output.Write(buf, 0, n);
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                try { if (Directory.Exists(destDir)) Directory.Delete(destDir, true); }
                catch (Exception) { /* 尽力而为 */ }
                if (ex is BridgeException) throw;
                throw new BridgeException("ZIP_UNSAFE_ENTRY", "zip could not be read: " + ex.Message);
            }
        }
    }
}
